core/src/com/unciv/ui/components/SortableGrid.kt

```diff
@@ -0,0 +1,212 @@
+@file:Suppress("MemberVisibilityCanBePrivate") // A generic Widget has public members not necessarily used elsewhere
+package com.unciv.ui.components
+
+import com.badlogic.gdx.graphics.Color
+import com.badlogic.gdx.scenes.scene2d.Actor
+import com.badlogic.gdx.scenes.scene2d.Group
+import com.badlogic.gdx.scenes.scene2d.ui.Cell
+import com.badlogic.gdx.scenes.scene2d.ui.Label
+import com.badlogic.gdx.scenes.scene2d.ui.Table
+import com.badlogic.gdx.utils.Align
+import com.unciv.ui.components.UncivTooltip.Companion.addTooltip
+import com.unciv.ui.components.extensions.*
+import com.unciv.ui.screens.basescreen.BaseScreen
+
+
+/**
+ * A generic sortable grid Widget
+ *
+ * @param IT Type of the data objects that provide info per row
+ * @param ST Type for [parentScreen], Anything allowed, specific meaning defined only by ISortableGridContentProvider subclass
+ * @param CT Type of the columns
+ */
+class SortableGrid<IT, ST, CT: ISortableGridContentProvider<IT, ST>> (
+    /** Provides the columns to render as [ISortableGridContentProvider] instances */
+    private val columns: Iterable<CT>,
+    /** Provides the actual "data" as in one object per row that can then be passed to [ISortableGridContentProvider] methods to fetch cell content */
+    private val data: Iterable<IT>,
+    /** Passed to [ISortableGridContentProvider.getEntryActor] where it can be used to define `onClick` actions. */
+    private val parentScreen: ST,
+    /** Sorting state will be kept here - provide your own e.g. if you want to persist it */
+    private val sortState: ISortState<CT> = SortState(columns.first()),
+    /** Size for header icons - if you set this too low, there is a chance that the tables will be misaligned */
+    private val iconSize: Float = 50f,
+    /** vertical padding for all Cells */
+    paddingVert: Float = 5f,
+    /** horizontal padding for all Cells */
+    paddingHorz: Float = 8f,
+    /** When `true`, the header row isn't part of the widget but delivered through [getHeader] */
+    private val separateHeader: Boolean = false,
+    /** Called after every update - during init and re-sort */
+    private val updateCallback: ((header: Table, details: Table, totals: Table) -> Unit)? = null
```

Comment:
You usually over-generalize things, so I'll say ahead of time that any of these that is currently *not* an input parameter - shouldn't be. We'll add them if needed.